### PR TITLE
Implement VPC Endpoint for AWS Provider

### LINKED ISSUE #715: VPC endpoints support for aws provider
Currently there is no support for vpc endpoints (s3, ecr kms etc). 
Clusters that reside in private/local subnets need a connection to aws services in order to pull/push docker images(ecr), use s3 or kms without going through the internet. 

v0.18.1

### PR DESCRIPTION
**Pull request by darryl-sw**
### Description of your changes

Fixes #715 #467

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

- Verified that the VPC Endpoint can be created upstream in AWS
- Ensured that resource is marked as READY only when VPC Endpoint status is available
- Verified that the VPC Endpoint can be deleted upstream in AWS

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: examples/ec2/vpcendpoint.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+apiVersion: ec2.aws.crossplane.io/v1alpha1
+kind: VPCEndpoint
+metadata:
+  name: sample-vpcendpoint
+spec:
+  forProvider:
+    region: us-east-1
+    privateDNSEnabled: true
+    serviceName: com.amazonaws.us-east-1.vpce-svc-0123456789abcdef0
```
Comment (chlunde) [inline examples/ec2/vpcendpoint.yaml:9]: is it possible to provide values here (serviceName, securityGroupIDs, subnetIDs, vpcID) that will work with the existing examples?

File: pkg/controller/ec2/vpcendpoint/setup.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	cr.Status.AtProvider.VPCEndpoint = generateVPCEndpointSDK(resp.VpcEndpoints[0])
+
+	switch awsclients.StringValue(resp.VpcEndpoints[0].State) {
+	case "available":
+		cr.SetConditions(xpv1.Available())
+	case "pending":
+		cr.SetConditions(xpv1.Creating())
+	case "pending-acceptance":
```
Comment (chlunde) [inline pkg/controller/ec2/vpcendpoint/setup.go:115]: looks like these are available as constants, so you may consider using them to avoid typos: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2/types#VpcPeeringConnectionStateReasonCode

but I don't see `pending` there? should it be `VpcPeeringConnectionStateReasonCodeProvisioning`?

you can also use `case CodeFoo, CodeBar:` when two states should map to the same condition.

File: pkg/controller/ec2/vpcendpoint/setup.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		cr.SetConditions(xpv1.Unavailable())
+	case "deleting":
+		cr.SetConditions(xpv1.Deleting())
+	}
+
+	return obs, nil
+}
+
+func isUpToDate(*svcapitypes.VPCEndpoint, *svcsdk.DescribeVpcEndpointsOutput) (bool, error) {
+	return true, nil
```
Comment (chlunde) [inline pkg/controller/ec2/vpcendpoint/setup.go:127]: can you implement this? what fields are mutable?

also check pre/post update in another controller, like 
https://github.com/crossplane/provider-aws/blob/29466b1812a013378ba7bf845d712000c302408a/pkg/controller/docdb/dbsubnetgroup/setup.go#L137-L150

File: pkg/controller/ec2/vpcendpoint/setup.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
+package vpcendpoint
```
Comment (chlunde) [inline pkg/controller/ec2/vpcendpoint/setup.go:1]: can you add unit tests for this file? look for files called `setup_test.go` for examples, like https://github.com/crossplane/provider-aws/blob/master/pkg/controller/docdb/dbsubnetgroup/setup_test.go

**Review by chlunde — CHANGES_REQUESTED** [review]
Thanks for your contribution, could you also look at bit at supporting updates?

File: pkg/controller/ec2/vpcendpoint/setup.go (hunk 2 above)
Comment (dare-darryl) [reply pkg/controller/ec2/vpcendpoint/setup.go:115]: It seems that those constants are for VPC Peering, and not for VPC Endpoints.

The closest constants that seems to match the state for VPC Endpoints are here: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2/types#State

However, the casing in the constants are wrong. Hence, they were hard-coded instead. "pending" is indeed the correct state to be used for creating.

I'll fix the rest of the items.

File: examples/ec2/vpcendpoint.yaml (hunk 1 above)
Comment (darryl-sw) [reply examples/ec2/vpcendpoint.yaml]: This has been added.

File: pkg/controller/ec2/vpcendpoint/setup.go (hunk 3 above)
Comment (darryl-sw) [reply pkg/controller/ec2/vpcendpoint/setup.go]: This is now implemented.

File: pkg/controller/ec2/vpcendpoint/setup.go (hunk 4 above)
Comment (darryl-sw) [reply pkg/controller/ec2/vpcendpoint/setup.go:1]: Unit tests are added.

**Comment by darryl-sw** [comment]
@chlunde I've completed the requested changes. Could you help me take a look at the changes please? Thanks!

File: pkg/controller/ec2/vpcendpoint/setup.go [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointInput) error {
+	// set external name as tag on the vpc endpoint
+	resType := "vpc-endpoint"
+	key := "Name"
+	value := meta.GetExternalName(cr)
+
+	// Clear SGs, RTs, and Subnets if they're empty
+	if cr.Spec.ForProvider.SecurityGroupIDs == nil || len(cr.Spec.ForProvider.SecurityGroupIDs) == 0 {
```
Comment (turkenh) [inline pkg/controller/ec2/vpcendpoint/setup.go:77]: I believe we could drop these nil checks: https://medium.com/@habibridho/golang-nil-vs-empty-slice-87fd51c0a4d

Comment (darryl-sw) [reply pkg/controller/ec2/vpcendpoint/setup.go]: Nil checks are dropped for this.

File: apis/ec2/v1alpha1/custom_types.go [hunk 6]
```diff
@@ -32,3 +32,44 @@ type CustomVPCPeeringConnectionParameters struct {
 	// connection if both VPCs are in the same tenant.
 	AcceptRequest bool `json:"acceptRequest,omitempty"`
 }
+
+// CustomVPCEndpointParameters are custom parameters for VPCEndpoint
+type CustomVPCEndpointParameters struct {
+	// The ID of the VPC. You must specify this parameter in the request.
```
[the remark this answers is not in the document]
Comment (darryl-sw) [reply apis/ec2/v1alpha1/custom_types.go:38]: This has been marked as `+optional`.

File: apis/ec2/v1alpha1/referencers.go [hunk 7]
```diff
@@ -60,3 +61,67 @@ func (mg *VPCPeeringConnection) ResolveReferences(ctx context.Context, c client.
 
 	return nil
 }
+
+// ResolveReferences of this VPCEndpoint
+func (mg *VPCEndpoint) ResolveReferences(ctx context.Context, c client.Reader) error {
```
[the remark this answers is not in the document]
Comment (darryl-sw) [reply apis/ec2/v1alpha1/referencers.go:66]: Hi @turkenh, unfortunately, I found difficulty in using the angryjet tool. I'm not too sure how I can make use of the tool.

Some of the areas in which I found difficulties:
1. Referencing the extractor: the example provides an extractor that makes use of v1beta1.SubnetARN(). However, in most cases, we make use of reference.ExternalName(). Should it be replaced 1:1 like that?
2. Unsure of how/where to run the angryjet tool. I ran `angryjet generate-methodsets github.com/crossplane/crossplane/apis` from the provider-aws directory and received the following errors: `angryjet: error: error loading packages using pattern github.com/crossplane/crossplane/apis: -: no required module provides package github.com/crossplane/crossplane/apis; to add it:
        go get github.com/crossplane/crossplane/apis`

Would you be able to guide me along for this, or can we delay the use of this generate feature till another time?

**Comment by darryl-sw** [comment]
@turkenh, I failed the check-diff job in the previous run. Did a `make generate` and `git diff` locally to get this output.

```
diff --git a/apis/acm/v1alpha1/zz_generated.deepcopy.go b/apis/acm/v1alpha1/zz_generated.deepcopy.go
index 36535911..cf8dfdbe 100644
--- a/apis/acm/v1alpha1/zz_generated.deepcopy.go
+++ b/apis/acm/v1alpha1/zz_generated.deepcopy.go
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated
 
 /*
diff --git a/apis/acmpca/v1alpha1/zz_generated.deepcopy.go b/apis/acmpca/v1alpha1/zz_generated.deepcopy.go
index 05d1ef6f..943bc892 100644
--- a/apis/acmpca/v1alpha1/zz_generated.deepcopy.go
+++ b/apis/acmpca/v1alpha1/zz_generated.deepcopy.go
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated
 
 /*
diff --git a/apis/apigatewayv2/v1alpha1/zz_generated.deepcopy.go b/apis/apigatewayv2/v1alpha1/zz_generated.deepcopy.go
index 6de8677b..bcc03079 100644
--- a/apis/apigatewayv2/v1alpha1/zz_generated.deepcopy.go
+++ b/apis/apigatewayv2/v1alpha1/zz_generated.deepcopy.go
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated
 
 /*
diff --git a/apis/cache/v1alpha1/zz_generated.deepcopy.go b/apis/cache/v1alpha1/zz_generated.deepcopy.go
index ca1083d6..3fb44a5e 100644
--- a/apis/cache/v1alpha1/zz_generated.deepcopy.go
+++ b/apis/cache/v1alpha1/zz_generated.deepcopy.go
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated
 
 /*
diff --git a/apis/cache/v1beta1/zz_generated.deepcopy.go b/apis/cache/v1beta1/zz_generated.deepcopy.go
index d671c164..3543222b 100644
--- a/apis/cache/v1beta1/zz_generated.deepcopy.go
+++ b/apis/cache/v1beta1/zz_generated.deepcopy.go
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated

 ...
// 
```

34 files came appended with `//go:build !ignore_autogenerated`. By any chance, would you happen to know how I can resolve this?

**Comment by chlunde** [comment]
> ```
> +++ b/apis/cache/v1beta1/zz_generated.deepcopy.go
> @@ -1,3 +1,4 @@
> +//go:build !ignore_autogenerated
>  // +build !ignore_autogenerated
> ```
> 34 files came appended with `//go:build !ignore_autogenerated`. By any chance, would you happen to know how I can resolve this?

this happens when you have go 1.17 locally, and the build environment uses go 1.16 or older. A workaround is currently something like `sed -i /go:build/ apis/*/*/*.go` (and more) before commit/push.

File: apis/ec2/v1alpha1/referencers.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply apis/ec2/v1alpha1/referencers.go:66]: @darryl-sw I wouldn't mean delaying this PR with this, given you already implemented and probably tested. So, I would leave the decision to you.

> 1. Referencing the extractor: the example provides an extractor that makes use of v1beta1.SubnetARN(). However, in most cases, we make use of reference.ExternalName(). Should it be replaced 1:1 like that?

I believe @muvaf could better answer this than me.

> 2. Unsure of how/where to run the angryjet tool. I ran angryjet generate-methodsets github.com/crossplane/crossplane/apis from the provider-aws directory and received the following errors: angryjet: error: error loading packages using pattern github.com/crossplane/crossplane/apis: -: no required module provides package github.com/crossplane/crossplane/apis; to add it: go get github.com/crossplane/crossplane/apis

I don't think you should do something special other than updating `crossplane/crossplane-tools` dependency to the latest master in go.mod.

[the remark this answers is not in the document]
Comment (darryl-sw) [reply apis/ec2/v1alpha1/referencers.go:66]: @turkenh in that case, I will leave it as is, and mark this conversation as resolved. Thank you!

File: pkg/controller/ec2/vpcendpoint/setup.go [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+	/*
+		4. Check policyDocument
+	*/
+	defaultPolicy := "{\"Statement\":[{\"Action\":\"*\",\"Effect\": \"Allow\",\"Principal\":\"*\",\"Resource\":\"*\"}]}"
+	declaredPolicy := awsclients.StringValue(cr.Spec.ForProvider.PolicyDocument)
+	upstreamPolicy := awsclients.StringValue(obj.VpcEndpoints[0].PolicyDocument)
+
+	// If no declared policy, we expect the result to be equivalent to the default policy
+	if declaredPolicy == "" {
+		difference, _ := jsondiff.Compare([]byte(upstreamPolicy), []byte(defaultPolicy), &jsondiff.Options{})
```
[the remark this answers is not in the document]
Comment (darryl-sw) [reply pkg/controller/ec2/vpcendpoint/setup.go:191]: @chlunde how would you recommend me resolving this?

[the remark this answers is not in the document]
Comment (chlunde) [reply pkg/controller/ec2/vpcendpoint/setup.go:191]: I'm moving the function so you can do `return awsclients.IsPolicyUpToDate(pol1, pol2), nil`

[the remark this answers is not in the document]
Comment (darryl-sw) [reply pkg/controller/ec2/vpcendpoint/setup.go]: This is resolved.

**Comment by darryl-sw** [comment]
@chlunde can you take a look at the changes please?

File: go.mod [hunk 9]
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/go-ini/ini v1.46.0
 	github.com/google/go-cmp v0.5.5
 	github.com/mitchellh/copystructure v1.0.0
+	github.com/nsf/jsondiff v0.0.0-20210303162244-6ea32392771e
```
Comment (chlunde) [inline go.mod:14]: Is this used now? maybe you can revert the changes to `go.mod` and `go.sum` now.

File: pkg/controller/ec2/vpcendpoint/setup.go [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+	}
+	if len(cr.Spec.ForProvider.RouteTableIDs) == 0 {
+		obj.RouteTableIds = nil
+	}
+	if len(cr.Spec.ForProvider.SubnetIDs) == 0 {
+		obj.SubnetIds = nil
+	}
+
+	// Set tags
+	spec := svcsdk.TagSpecification{
```
Comment (chlunde) [inline pkg/controller/ec2/vpcendpoint/setup.go:87]: this should be a loop like this?
```go
import	"github.com/crossplane/crossplane-runtime/pkg/resource"

	for k, v := range resource.GetExternalTags(mg) {
```

it is typically done via `SetupFoo`:
```
			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), &tagger{kube: mgr.GetClient()}),
```


https://github.com/crossplane/provider-aws/blob/3f1b7187455f14c306b3037312a19a791ac9d9eb/pkg/controller/ec2/address/controller.go#L222-L248

but I have to say that is a lot of duplicate code, so maybe we should discuss a generic `type Tagger interface { SetTag(k, v string) GetTag(k string) }` with @muvaf etc.

File: package/crds/ec2.aws.crossplane.io_subnets.yaml [hunk 11]
```diff
@@ -28,6 +28,9 @@ spec:
             type: string
           availabilityZoneID:
             type: string
+          availableIPAddressCount:
```
[the remark this answers is not in the document]
Comment (darryl-sw) [reply package/crds/ec2.aws.crossplane.io_subnets.yaml:31]: This was auto-generated, I'm not making use of it, nor do I really know what this is for. Should I revert this change?

File: pkg/controller/ec2/vpcendpoint/setup.go (hunk 10 above)
Comment (darryl-sw) [reply pkg/controller/ec2/vpcendpoint/setup.go]: This is resolved. Tagging is moved to using an Initializer.

File: go.mod (hunk 9 above)
Comment (darryl-sw) [reply go.mod]: Package has been removed.

File: package/crds/ec2.aws.crossplane.io_subnets.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (darryl-sw) [reply package/crds/ec2.aws.crossplane.io_subnets.yaml:31]: @chlunde following up on this.

[the remark this answers is not in the document]
Comment (darryl-sw) [reply package/crds/ec2.aws.crossplane.io_subnets.yaml:31]: Hi @chlunde, may I check how this should be resolved or addressed please?

[the remark this answers is not in the document]
Comment (chlunde) [reply package/crds/ec2.aws.crossplane.io_subnets.yaml:31]: I haven't seen this issue anywhere else, not sure why it pops up now. This seems to fix it:
```diff
diff --git a/apis/ec2/v1alpha1/generator-config.yaml b/apis/ec2/v1alpha1/generator-config.yaml
index 7b1eaa92..cf2ae446 100644
--- a/apis/ec2/v1alpha1/generator-config.yaml
+++ b/apis/ec2/v1alpha1/generator-config.yaml
@@ -71,3 +71,4 @@ ignore:
     - CreateVpcEndpointInput.VpcId
     - ModifyVpcEndpointInput.VpcId
     - DeleteVpcEndpointInput.VpcId
+    - Subnet.AvailableIpAddressCount
```

Probably it has not been detected because `ec2` is missing from the Makefile:
```diff
diff --git a/Makefile b/Makefile
index 920f543c..2f6af424 100644
--- a/Makefile
+++ b/Makefile
@@ -7,7 +7,7 @@ PROJECT_REPO := github.com/crossplane/$(PROJECT_NAME)
 PLATFORMS ?= linux_amd64 linux_arm64
 
 CODE_GENERATOR_COMMIT ?= cac5654b7bb64c8f754ad9af01799ef70d9541b6
-GENERATED_SERVICES="apigatewayv2,cloudfront,dynamodb,efs,kafka,kms,lambda,rds,secretsmanager,servicediscovery,sfn"
+GENERATED_SERVICES="apigatewayv2,cloudfront,dynamodb,efs,ec2,kafka,kms,lambda,rds,secretsmanager,servicediscovery,sfn"
 
 # kind-related versions
 KIND_VERSION ?= v0.11.1
```

**Comment by chlunde** [comment]
@AaronME is there an issue in detect-noop?

@darryl-sw The tests are not running, when I run them locally I get an issue with a panic in `IsPolicyUpToDate`. I think we should harden it by using `StringValue`:

```diff
diff --git a/pkg/clients/aws.go b/pkg/clients/aws.go
index 7cd91a7f..83a8b35e 100644
--- a/pkg/clients/aws.go
+++ b/pkg/clients/aws.go
@@ -658,11 +658,11 @@ func IsPolicyUpToDate(local, remote *string) bool {
        var remoteUnmarshalled interface{}
 
        var err error
-       err = json.Unmarshal([]byte(*local), &localUnmarshalled)
+       err = json.Unmarshal([]byte(StringValue(local)), &localUnmarshalled)
        if err != nil {
                return false
        }
-       err = json.Unmarshal([]byte(*remote), &remoteUnmarshalled)
+       err = json.Unmarshal([]byte(StringValue(remote)), &remoteUnmarshalled)
        if err != nil {
                return false
        }
```

@darryl-sw I don't think you have run code gen, I get a diff here, could you run `make generate`?

```diff
diff --git a/pkg/controller/ec2/vpcendpoint/zz_conversions.go b/pkg/controller/ec2/vpcendpoint/zz_conversions.go
index 2be39997..e52368a3 100644
--- a/pkg/controller/ec2/vpcendpoint/zz_conversions.go
+++ b/pkg/controller/ec2/vpcendpoint/zz_conversions.go
@@ -88,11 +88,6 @@ func GenerateVPCEndpoint(resp *svcsdk.DescribeVpcEndpointsOutput) *svcapitypes.V
                } else {
                        cr.Spec.ForProvider.VPCEndpointType = nil
                }
-               if elem.VpcId != nil {
-                       cr.Spec.ForProvider.VPCID = elem.VpcId
-               } else {
-                       cr.Spec.ForProvider.VPCID = nil
-               }
                found = true
                break
        }
@@ -177,9 +172,6 @@ func GenerateCreateVpcEndpointInput(cr *svcapitypes.VPCEndpoint) *svcsdk.CreateV
        if cr.Spec.ForProvider.VPCEndpointType != nil {
                res.SetVpcEndpointType(*cr.Spec.ForProvider.VPCEndpointType)
        }
-       if cr.Spec.ForProvider.VPCID != nil {
-               res.SetVpcId(*cr.Spec.ForProvider.VPCID)
-       }
 
        return res
 }
```

File: package/crds/ec2.aws.crossplane.io_subnets.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (darryl-sw) [reply package/crds/ec2.aws.crossplane.io_subnets.yaml]: @chlunde I've resolved this following your input. Thanks!

**Comment by darryl-sw** [comment]
@chlunde I've made the requested changes to `isPolicyUpToDate` and reran `make generate`.

**Review by chlunde — APPROVED** [review]

**Comment by muvaf** [comment]
@darryl-sw seems like there are some conflicts need to be resolved. Could you take care of those so that we can merge?

**Comment by haarchri** [comment]
@darryl-sw can you rebase ? And squash commits ? After that we can merge the feature - thanks

**Comment by darryl-sw** [comment]
@haarchri running `make generate` with "ec2" added to `GENERATED_SERVICES` in the `Makefile` results in an error that I have not been able to resolve after rebasing.

```
13:15:38 [ OK ] go modules dependencies verified
13:15:39 [ .. ] go generate darwin_amd64
github.com/crossplane/provider-aws/apis/ec2/v1alpha1:-: CRD for Instance.ec2.aws.crossplane.io has no storage version
Error: not all generators ran successfully
run `controller-gen object:headerFile=../hack/boilerplate.go.txt paths=./... crd:trivialVersions=true,allowDangerousTypes=true,crdVersions=v1 output:artifacts:config=../package/crds -w` to see all available markers, or `controller-gen object:headerFile=../hack/boilerplate.go.txt paths=./... crd:trivialVersions=true,allowDangerousTypes=true,crdVersions=v1 output:artifacts:config=../package/crds -h` for usage
exit status 1
apis/generate.go:26: running "go": exit status 1
```

I took a look at the file `package/crds/ec2.aws.crossplane.io_instances.yaml` and identified that there were two versions, both named `v1alpha1`, both of them with `storage: false`.

To further isolate the problem, I verified that simply adding "ec2" to `GENERATED_SERVICES` on the master branch and running `make generate` produces the same error message.

Would you, or anybody, be able to advice on how I can fix this please?

**Comment by haarchri** [comment]
@darryl-sw sorry i missed that we are blocked with ec2 services because of https://github.com/crossplane/provider-aws/issues/876

**Comment by darryl-sw** [comment]
@haarchri shall I wait for #876 first before proceeding with the rebase then?

**Comment by hanlins** [comment]
Any updates on the PR? Blockers? cc @haarchri

**Comment by haarchri** [comment]
@darryl-sw you can go forward the blockers are solved in master branch - thanks

**Comment by xvzf** [comment]
@haarchri what's the ETA for this to be released? :)

**Comment by haarchri** [comment]
@darryl-sw  can rebase and i can have a look - in the next days we will cut next release - so i have time for review since rebase is done

**Comment by darryl-sw** [comment]
@haarchri rebased and completed another round of upstream testing. :)

**Review by AaronME — APPROVED** [review]
Thanks @darryl-sw  for your contribution. I was able to create and delete a VPCEndpoint using this.

@haarchri this is good to rebase and merge.

**Comment by haarchri** [comment]
i will rebase now and adjust generated resolvers to get this PR merged ;)

**Comment by haarchri** [comment]
@AaronME done!
can you check build submodule ? looks like we see changes in this PR - after that we are ready to merge =)

tested also on my side:
```
NAME                                                   READY   SYNCED   EXTERNAL-NAME
vpcendpoint.ec2.aws.crossplane.io/sample-vpcendpoint   True    True     vpce-0037d4f0338c77881
```
